packages/core/saved-objects/core-saved-objects-base-server-internal/src/model_version/build_transform_fn.test.ts

```diff
@@ -9,119 +9,318 @@
 
-    const context = createContext();
-    const document = createDoc();
+      const context = createContext();
+      const document = createDoc();
 
-    const output = mergedTransform(document, context);
-    // test that it returns the source doc unmodified
-    expect(output.document).toEqual(createDoc());
+      mergedTransform(document, context);
+
+      expect(transform1).toHaveBeenCalledTimes(1);
+      expect(transform1).toHaveBeenCalledWith(document, context);
+      expect(transform2).toHaveBeenCalledTimes(1);
+      expect(transform2).toHaveBeenCalledWith(document, context);
+      expect(transform1.mock.invocationCallOrder[0]).toBeLessThan(
+        transform2.mock.invocationCallOrder[0]
+      );
+    });
+
+    it('executes many functions in order', () => {
+      const transform1 = stubDatabackfill();
+      const transform2 = stubDatabackfill();
+      const transform3 = stubDatabackfill();
+      const transform4 = stubDatabackfill();
+      const changes: SavedObjectsModelChange[] = [
+        { type: 'data_backfill', backfillFn: transform1 },
+        { type: 'data_backfill', backfillFn: transform2 },
+        { type: 'data_backfill', backfillFn: transform3 },
+        { type: 'data_backfill', backfillFn: transform4 },
+      ];
+
+      const mergedTransform = buildModelVersionTransformFn(changes);
+
+      const context = createContext();
+      const document = createDoc();
+
+      mergedTransform(document, context);
+
+      expect(transform1).toHaveBeenCalledTimes(1);
+      expect(transform2).toHaveBeenCalledTimes(1);
+      expect(transform3).toHaveBeenCalledTimes(1);
+      expect(transform4).toHaveBeenCalledTimes(1);
+
+      expect(transform1).toHaveBeenCalledWith(document, context);
+      expect(transform2).toHaveBeenCalledWith(document, context);
+      expect(transform3).toHaveBeenCalledWith(document, context);
+      expect(transform4).toHaveBeenCalledWith(document, context);
+
+      expect(transform1.mock.invocationCallOrder[0]).toBeLessThan(
+        transform2.mock.invocationCallOrder[0]
+      );
+      expect(transform2.mock.invocationCallOrder[0]).toBeLessThan(
+        transform3.mock.invocationCallOrder[0]
+      );
+      expect(transform3.mock.invocationCallOrder[0]).toBeLessThan(
+        transform4.mock.invocationCallOrder[0]
+      );
+    });
+  });
+
+  describe('converting mixed change types', () => {
+    it('chains and executes the changes', () => {
+      const changes: SavedObjectsModelChange[] = [
+        {
+          type: 'data_backfill',
+          backfillFn: (doc) => {
+            return { attributes: { newField: 'newValue' } };
+          },
+        },
+        {
+          type: 'data_removal',
+          removedAttributePaths: ['oldField'],
+        },
+        {
+          type: 'unsafe_transform',
+          transformFn: (document) => {
+            document.attributes.unsafeNewProp = 'unsafeNewValue';
+            return { document };
+          },
```

Comment:
I couldn't come up with a good idea to achieve this: it'd be nice to have TS checking the shape of the object that is arriving here rather than relying on developers to know/provide it.

Not the biggest issue it seems fairly trivial to create tests for these change chains, but this is something I don't think is easy/possible with current API and that safety rather comes from limiting capabilities.

Reply:
The functions are generics (inputAttributes, outputAttributes). The problem is that when defining the migration map, your can't specify generic types on the fly, e.g 

```
transformFn: (document) => {
            document.attributes.unsafeNewProp = 'unsafeNewValue';
            return { document };
          }
```

will use the default type, and it's very tedious to manually defines it, e.g

transformFn: (document) => {
            document.attributes.unsafeNewProp = 'unsafeNewValue';
            return { document };
          } as SOMVBlablaFunction<InputAttrs, OutputAttrs>

(and it also requires to import the types, where the fully inline def doesn't need it)

So we already have TS check support, it's just that the way TS works make it terrible to leverage it there...

Reply:
Yeah true, I guess I meant specifically having TS infer the type.